lib/web_ui/lib/src/engine/shadow.dart

```diff
@@ -5,366 +5,134 @@
-      ));
+  // tangent for x
+  final double tx = (kLightRadius + shape.width * 0.5) / kLightHeight;
+  // tangent for y
+  final double ty = (kLightRadius + shape.height * 0.5) / kLightHeight;
+  final double dx = elevation * tx;
+  final double dy = elevation * ty;
+  final ui.Offset offset = computeShadowOffset(elevation);
+  return ui.Rect.fromLTRB(
+    shape.left - dx,
+    shape.top - dy,
+    shape.right + dx,
+    shape.bottom + dy,
+  ).shift(offset);
+}
 
-      result.add(CanvasShadow(
-        offsetX: 0.0,
-        offsetY: 0.0,
-        blur: 5.0,
-        spread: -5.0,
-        color: umbraColor,
-      ));
-    } else {
-      result.add(CanvasShadow(
-        offsetX: 0.0,
-        offsetY: 24.0,
-        blur: 18.0,
-        spread: 3.0,
-        color: penumbraColor,
-      ));
+/// Information needed to render a shadow using CSS or canvas.
+@immutable
+class SurfaceShadowData {
+  const SurfaceShadowData({
+    @required this.blurWidth,
+    @required this.offset,
+  });
 
-      result.add(CanvasShadow(
-        offsetX: 0.0,
-        offsetY: 9.0,
-        blur: 23.0,
-        spread: 8.0,
-        color: ambientShadowColor,
-      ));
+  /// The length in pixels of the shadow.
+  ///
+  /// This is different from the `sigma` used by blur filters. This value
+  /// contains the entire shadow, so, for example, to compute the shadow
+  /// bounds it is sufficient to add this value to the width of the shape
+  /// that casts it.
+  final double blurWidth;
+
+  /// The offset of the shadow relative to the shape as computed by
+  /// [computeShadowOffset].
+  final ui.Offset offset;
+}
 
-      result.add(CanvasShadow(
-        offsetX: 0.0,
-        offsetY: 11.0,
-        blur: 7.5,
-        spread: -7.0,
-        color: umbraColor,
-      ));
-    }
-    return result;
+/// Computes the shadow for [shape] based on its [elevation] from the surface
+/// of the screen.
+///
+/// The algorithm approximates the math done by the C++ implementation from
+/// `physical_shape_layer.cc` but it's not exact, since on the Web we do not
+/// (cannot) use Skia's shadow API directly. However, this algorithms is
+/// consistent with [computePenumbraBounds] used by [RecordingCanvas] during
+/// bounds estimation.
+SurfaceShadowData computeShadow(ui.Rect shape, double elevation) {
+  if (elevation == 0.0) {
+    return null;
   }
-}
 
-class CanvasShadow {
-  CanvasShadow({
-    @required this.offsetX,
-    @required this.offsetY,
-    @required this.blur,
-    @required this.spread,
-    @required this.color,
-  });
+  final double penumbraTangentX =
+      (kLightRadius + shape.width * 0.5) / kLightHeight;
+  final double penumbraTangentY =
+      (kLightRadius + shape.height * 0.5) / kLightHeight;
+  final double penumbraWidth = elevation * penumbraTangentX;
+  final double penumbraHeight = elevation * penumbraTangentY;
+  return SurfaceShadowData(
+    // There's no way to express different blur along different dimensions, so
```

Comment:
along different dimensions in CSS, so...  To prevent the shadow blur from being longer than shape itself, using min instead of average of penumbra values.

Reply:
Done.